### DISCUSSION TITLE
How to wait for `vim.loop.spawn` running and get all stdout/stderr and exit the loop?

### DISCUSSION THREAD

**Discussion by linrongbin16** [neovim/neovim#25358] [Q&A] [answered]
Hi, I'm trying to run a command in Neovim with `vim.loop.spawn` and capture all the outputs:

```bash
rg --column --color=always
```

For now I'm trying to write it as:

```lua
local function run_async_cmd(cmds, data_consumer)
    local out_pipe = vim.loop.new_pipe(false) --[[@as uv_pipe_t]]
    local err_pipe = vim.loop.new_pipe(false) --[[@as uv_pipe_t]]

    local function on_exit()
        out_pipe:shutdown(function(shutdown_err)
            out_pipe:close(function(close_err)
                vim.loop.stop()
            end)
         end)
        err_pipe:shutdown(function(shutdown_err)
            err_pipe:close(function(close_err)
                if out_pipe():is_closing() then
                    vim.loop.stop()
                end
            end)
        end)
    end

    local process_handler, process_id = vim.loop.spawn(cmds[1], {
        args = { unpack(cmds, 2) },
        stdio = { nil, out_pipe, err_pipe },
    }, function(code, signal)
        out_pipe:read_stop()
        err_pipe:read_stop()
        on_exit()
    end)

    local function on_out(err, data)
        if err then
            on_exit()
            return
        end
        if not data then
            on_exit()
            return
        end
        data_consumer(data)
    end

    local function on_err(err, data)
        on_exit()
    end

    out_pipe:read_start(on_out)
    err_pipe:read_start(on_err)
    vim.loop.run()
end

```

So I could use it like:

```lua
run_async_cmd({'rg', '--column', '--color=always'}, function(data) print(data) end)
```

But somehow, when I use this function in my script, it's stucking and never been callback in my `data_consumer` function.

How should I start this function and wait for it done?

Note: I have tried `vim.loop.run()` and `vim.loop.stop()` when start running the loop, and stop the loop on exit.

**Comment by clason** [comment, marked as answer]
You could look at `vim.system` (on nightly).

**Reply by linrongbin16** [reply]
Hi @clason , I want to accept your sugguestion, but I'm writing a nvim plugin, which supports v0.5.0 (with `vim.loop`), I believe the support still makes sense to some one who still use v0.5.0.

So I want to stick to the API `vim.loop`.

**Reply by clason** [reply]
_Nobody_ should be using 0.5.0 anymore -- we only support the latest release (0.9.2) and don't have any sort of LTS policy (yet).

In any case, I just wanted to give a hint to look at the (code of) `vim.system` to show how to work with async `vim.loop.spawn`.

**Reply by linrongbin16** [reply]
Oh, sure, I could take a look at the source code of `vim.system` thanks !
